### PR TITLE
Make fetch() zone-aware without triggering extra requests or uncatchable errors.

### PR DESCRIPTION
**Pull request by MichaelBuerge**
Previous implementation called fetch() during setup in order to obtain a result promise from it so its constructor could be patched to make it zone-aware.
This had two problems:
- Calling fetch() would trigger an unwanted (and unnecessary) HTTP-request.
- Chrome threw an uncatchable error, which means with open dev tools will halt execution and enter the debugger (which is hugely annoying, as it happens on every page load).

This change replaces fetch() with a wrapper function that lazily ensures the result returned by fetch() is zone-aware.

### CODE REVIEW

**Comment by MichaelBuerge** [comment]
CLA signed.

**Comment by MichaelBuerge** [comment]
#446

**Comment by JiaLiPassion** [comment]
@MichaelBuerge , please read [DEVELOPER.md](https://github.com/angular/zone.js/blob/master/DEVELOPER.md), before commit part, so you can check your commit. Currently it seems you build failed with tslint part.

**Comment by MichaelBuerge** [comment]
@JiaLiPassion Apologies for missing the linter. Everything should be fine now (`npm run ci` goes through).

**Comment by JiaLiPassion** [comment]
@MichaelBuerge , you are welcome, could you check my review comment? especially whether the first promise is patched or not one.

**Comment by MichaelBuerge** [comment]
@JiaLiPassion I can't find your review comment. As of now, the pull request page still says "no reviews". (Am I looking in the wrong place?)

However, guessing from your comment here, your question seems to be whether the first time the patched fetch() function is called its result promise is properly patched.
That is indeed the case, as the constructor's prototype is modified before any caller can get a hold of the returned promise. Also, the fetch tests (in `test/common/Promise.spec.ts`) would not pass with an unpatched result promise.
Does this address your concern?

**Comment by JiaLiPassion** [comment]
@MichaelBuerge, sorry I forget to push submit review button...

> That is indeed the case, as the constructor's prototype is modified before any caller can get a hold of the returned promise. Also, the fetch tests (in test/common/Promise.spec.ts) would not pass with an unpatched result promise.
> Does this address your concern?

yeah, you are right! thank you for explaining this, I didn't realize that.

File: lib/zone.ts [hunk 1]
```diff
@@ -1313,38 +1313,47 @@ const Zone: ZoneType = (function(global: any) {
...
-      const nativePromise = this;
-      return new ZoneAwarePromise((resolve, reject) => {
-               NativePromiseThen.call(nativePromise, resolve, reject);
-             })
-          .then(onResolve, onReject);
+
+  const symbolThenPatched = __symbol__('thenPatched');
+
+  function patchThen(Ctor) {
+    let proto = Ctor.prototype;
```
Comment (MichaelBuerge) [inline lib/zone.ts]: done

File: lib/zone.ts [hunk 2]
```diff
@@ -1313,38 +1313,47 @@ const Zone: ZoneType = (function(global: any) {
...
+    let originalThen = proto.then;
+    // Keep a reference to the original method.
+    proto[symbolThen] = originalThen;
+
+    Ctor.prototype.then = function(onResolve, onReject) {
+      let wrapped = new ZoneAwarePromise(
```
Comment (MichaelBuerge) [inline lib/zone.ts]: done

**Comment by MichaelBuerge** [comment]
My pleasure!
I assume there's a process in place, but if there's anything left do so this gets merged eventually and the associated issue (https://github.com/angular/zone.js/issues/446) closed please let me know.

**Comment by MichaelBuerge** [comment]
Also addresses #476

**Pull request merged**